File: modules/react-native-turbo-lab/src/specs/NativeTurboLab.ts
import type { TurboModule } from 'react-native'
import { TurboModuleRegistry } from 'react-native'

/**
 * "모듈이 언제 만들어지는가" 를 관찰하기 위한 TurboModule.
 *
 * 핵심은 `getCreatedAtMs` 다. 이 값이 0 에 가까우면 앱 시작 때 만들어진 것이고,
 * 버튼을 누른 시점과 비슷하면 **그때 처음 만들어진 것**이다.
 */
export interface Spec extends TurboModule {
  /** 네이티브 라이브러리 로드 후 이 모듈 인스턴스가 생성되기까지 걸린 ms */
  readonly getCreatedAtMs: () => number
  /** 지금이 로드 후 몇 ms 인지 */
  readonly getUptimeMs: () => number
  /** 호출할 때마다 1 증가 */
  readonly ping: () => number
  /** 지금까지 ping 이 몇 번 불렸는지 */
  readonly getPingCount: () => number
}

export default TurboModuleRegistry.getEnforcing<Spec>('NativeTurboLab')

File: modules/react-native-turbo-lab/cpp/NativeTurboLab.cpp
#include "NativeTurboLab.h"

#include <chrono>

namespace facebook::react {

namespace {

// 네이티브 라이브러리가 로드될 때(= 앱 시작 무렵) 한 번 초기화된다.
// 이후 모든 시각을 이 시점 기준으로 잰다.
const std::chrono::steady_clock::time_point kLoadTime =
    std::chrono::steady_clock::now();

double msSinceLoad() {
  const auto delta = std::chrono::steady_clock::now() - kLoadTime;
  return std::chrono::duration<double, std::milli>(delta).count();
}

} // namespace

// ★ 이 생성자가 언제 불리는지가 이 레슨의 핵심.
//   앱 시작 때가 아니라, JS 가 이 모듈을 처음 찾을 때 불린다.
NativeTurboLab::NativeTurboLab(std::shared_ptr<CallInvoker> jsInvoker)
    : NativeTurboLabCxxSpec(std::move(jsInvoker)),
      createdAtMs_(msSinceLoad()),
      pingCount_(0) {}

double NativeTurboLab::getCreatedAtMs(jsi::Runtime&) {
  return createdAtMs_;
}

double NativeTurboLab::getUptimeMs(jsi::Runtime&) {
  return msSinceLoad();
}

double NativeTurboLab::ping(jsi::Runtime&) {
  pingCount_ += 1;
  return pingCount_;
}

double NativeTurboLab::getPingCount(jsi::Runtime&) {
  return pingCount_;
}

// ⑧ 벤치마크용. Nitro · raw JSI 의 add 와 같은 일을 한다.
double NativeTurboLab::add(jsi::Runtime&, double a, double b) {
  return a + b;
}

} // namespace facebook::react
